samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java

```diff
@@ -0,0 +1,184 @@
+   * Array fields are converted from Avro {@link org.apache.avro.generic.GenericData.Array} to {@link ArrayList}
+   */
+  public enum AvroToRelObjConverter {
+
+    /**
+     * If the relational field type is ArraySqlType, We expect the avro field to be of type either
+     * {@link GenericData.Array} or {@link List} which then is converted to Rel field of type {@link ArrayList}
+     */
+    ArraySqlType {
+      @Override
+      Object convert(Object avroObj) {
+        ArrayList<Object> retVal = new ArrayList<>();
+        if (avroObj != null) {
+          if (avroObj instanceof GenericData.Array) {
+            retVal.addAll(((GenericData.Array) avroObj));
+          } else if (avroObj instanceof List) {
+            retVal.addAll((List) avroObj);
+          }
+        }
+
+        return retVal;
+      }
+    },
+
+    /**
+     * If the relational field type is MapSqlType, We expect the avro field to be of type
+     * {@link Map}
+     */
+    MapSqlType {
+      @Override
+      Object convert(Object obj) {
+        Map<String, Object> retVal = new HashMap<>();
+        if (obj != null) {
+          retVal.putAll((Map<String, ?>) obj);
+        }
+        return retVal;
+      }
+    },
+
+    /**
+     * If the relational field type is RelRecordType, The field is considered an object
+     * and moved to rel field without any translation.
+     */
+    RelRecordType {
+      @Override
+      Object convert(Object obj) {
+        return obj;
+      }
+    },
+
+    /**
+     * If the relational field type is BasicSqlType, The field is moved to rel field without any translation.
+     */
+    BasicSqlType {
+      @Override
+      Object convert(Object obj) {
+        return obj;
+      }
+    };
+
+    abstract Object convert(Object obj);
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroRelConverter.class);
+
+  private final Schema arraySchema = Schema.parse(
+      "{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Object\",\"namespace\":\"java.lang\",\"fields\":[]},\"java-class\":\"java.util.List\"}");
+  private final Schema mapSchema = Schema.parse(
+      "{\"type\":\"map\",\"values\":{\"type\":\"record\",\"name\":\"Object\",\"namespace\":\"java.lang\",\"fields\":[]}}");
+
+
+  public AvroRelConverter(SystemStream systemStream, AvroRelSchemaProvider schemaProvider, Config config) {
+    this.config = config;
+    this.relationalSchema = schemaProvider.getRelationalSchema();
+    this.avroSchema = Schema.parse(schemaProvider.getSchema(systemStream));
+  }
+
+  @Override
+  public SamzaSqlRelMessage convertToRelMessage(KV<Object, Object> samzaMessage) {
+    List<Object> values = new ArrayList<>();
+    List<String> fieldNames = new ArrayList<>();
+    Object value = samzaMessage.getValue();
+    if (value instanceof IndexedRecord) {
+      IndexedRecord record = (IndexedRecord) value;
+      fieldNames.addAll(relationalSchema.getFieldNames());
+      Schema recordSchema = record.getSchema();
```

Comment:
Great Catch!! Changed it to use the schema on the system.stream, which is probably right thing to do since the calcite plan was created using that schema.

*Reason why i was using the schema from the event*
The schema associated to the system.stream is the latest schema on that stream. The schema on the record is the schema associated to that record, They may not be same as the schema on stream, Hence i was using the schema on the record. But we require the schema on the system.stream to be compatible with all the events in the stream, So your recommendation makes sense.